examples/Fine_tuning_for_function_calling.ipynb

```diff
@@ -0,0 +1,1274 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "challenging_prompts = ['Play pre-recorded audio message',\n",
+    " 'Initiate live-streaming on social media',\n",
+    " 'Scan environment for heat signatures',\n",
+    " 'Enable stealth mode',\n",
+    " \"Change drone's paint job color\"]\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Play pre-recorded audio message \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"reject_request\",\n",
+      "  \"arguments\": \"{}\"\n",
+      "}\n",
+      "Initiate live-streaming on social media \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"control_camera\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"video\\\",\\n  \\\"duration\\\": 0\\n}\"\n",
+      "}\n",
+      "Scan environment for heat signatures \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"control_camera\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"photo\\\"\\n}\"\n",
+      "}\n",
+      "Enable stealth mode \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"set_drone_lighting\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"off\\\"\\n}\"\n",
+      "}\n",
+      "Change drone's paint job color \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"configure_led_display\",\n",
+      "  \"arguments\": \"{\\n  \\\"pattern\\\": \\\"solid\\\",\\n  \\\"color\\\": \\\"blue\\\"\\n}\"\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "for prompt in challenging_prompts:\n",
+    "  messages = []\n",
+    "  messages.append({\"role\": \"system\", \"content\": DRONE_SYSTEM_PROMPT})\n",
+    "  messages.append({\"role\": \"user\", \"content\": prompt})\n",
+    "  completion = get_function_call(model=\"gpt-3.5-turbo\",messages=messages,functions=function_list)\n",
+    "  print(prompt,'\\n')\n",
+    "  try:\n",
+    "    print(completion.function_call)\n",
+    "  except:\n",
+    "    print(completion.content)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we run into some problems.\n",
+    "The model here should reject all of these requests, as they are impossible given the functions, however instead the model calls functions that are somewhat related to the request, but incorrect. The model sets the camera to video when asked to begin 'live streaming to social media', and changes the LED's to blue when asked to 'change the paint color'... \\\n",
+    "Even with prompt engineering (in the prompt we say to decline requests when there is any ambiguity or doubt), the function calling is not working well. So this is a good opportuntiy for fine tuning for function calling."
```

Comment:
I'm not fully convinced by this, it feels to me like we just need better prompt engineering here. But maybe for the purpose of the example we can just say something like "in this simple case, these issues might be solvable through prompt engineering, but for the purpose of the example we will demonstrate how fine-tuning can be used instead"